### ISSUE TITLE
BN in VAE-GAN

### ISSUE THREAD

**Issue by jessejchuang** [2017-fall-DL-training-program/VAE-GAN-and-VAE-GAN#17]
Hi TA,

Take the following Encoder for an example, **_the BN output always got zeros_**. Shall we initialize the BN parameters before training?

- **self.bn1 = BatchNorm1d(2048)**
- **Input to self.bn1**
 0.0038  0.0481 -0.0533  ...  -0.0181  0.0303  0.0629
[torch.FloatTensor of size 1x2048]
- **Ouput from self.bn1**
    0     0     0  ...      0     0     0
[torch.FloatTensor of size 1x2048]
- **self.bn1.weight**
Parameter containing:
 0.7933
 0.4399
 0.1002
   ...   
 0.9673
 0.6917
 0.2108
[torch.FloatTensor of size 2048]
- **self.bn1.bias**
Parameter containing:
 0
 0
 0
...
 0
 0
 0
[torch.FloatTensor of size 2048]

**Comment by hui-po-wang** [comment]
Hi @jessejchuang ,

The algorithm of batch normalization is shown as below. You can see that after applying normalization to the original x and getting x head as the normalized result, the BN algorithm also operates an affine transform on x head to compute y. If gamma and beta are both equal to zero, it means this BN layer "only" normalize this tensor. However if either beta is not zero or gamma is not an identity matrix, you can regard the BN layer as an additional non-linearity operation so that the BN layer can increase the capacity of the model.

In conclusion, it's fine that bias equals zero and weights equal an identity matrix.

![default](https://user-images.githubusercontent.com/10228478/33648843-a1adb596-da96-11e7-9092-32ffab3bf9fa.png)

Thanks

**Comment by jessejchuang** [comment]
Hi TA,

Thanks for the detail explanation. I finally found it's the Pytorch Windows's issue. The same BN code works well in Linux. Since Pytorch Windows is not an official release, let's leave this question alone.
